File: prison_escape.pl
:- use_module(library(ansi_term)).

lokacja(biblioteka, 'Jesteś w bibliotece.').
lokacja(cela, 'Jesteś w swojej celi.').
lokacja(pralnia, 'Jesteś w pralni.').
lokacja(spacerniak, 'Jesteś na spacerniaku.').
lokacja(stolowka, 'Jesteś na więziennej stołówce.').

:- dynamic ekwipunek/1, kwestia/2, postac_w/2, pozycja_gracza/1, przedmiot_w/2.

akcja_uzycia(pamietnik) :-
    Do_ucieczki = [
        'broń (ze skarpetki i mydła)',
        'drut (z odkurzacza)',
        'farba',
        'kontakt poza więzieniem',
        'materiał (wyciąć z piżamy)',
        'papier',
        'płaszcze przeciwdeszczorwe',
        'sztućce',
        'sznurek (wyciąć z piżamy)',
        'śrubokręt',
        'ubrania na zmiane',
        'włosy (obciąć własne)'
    ],
    write('Potrzebne do ucieczki:'), nl,
    forall(member(Tekst, Do_ucieczki),
      (write('- '), write(Tekst), nl)), !.

akcja_uzycia(nozyczki) :-
    \+ pozycja_gracza(cela),
    ansi_format([fg(red)], 'Chcesz obciąć włosy? Mądre... ale zrób to w celi, żeby nikt się nie zainteresował.~n', []), !.

akcja_uzycia(nozyczki) :-
    pozycja_gracza(cela),
    \+ ekwipunek(wlosy),
    assertz(ekwipunek(wlosy)),
    ansi_format([fg(green)], 'Obciąłeś włosy.~n', []), !.

% daj(red, atlas) :-
%     ekwipunek(atlas),
%     retract(ekwipunek(atlas)),
%     retractall(kwestia(red, _)),
%     assertz(ekwipunek(kontakt)),
%     assertz(kwestia(red, 'Super, dzięki za atlas! Sprawę kontaktu możesz uznać za załatwioną.')).

dialog(Postac, Tekst) :-
    ansi_format([fg(cyan), bold], '~w: ', [Postac]),
    ansi_format([fg(white)], '~w~n', [Tekst]), !.

dialog_postaci(Postac) :-
    kwestia(Postac, Tekst),
    dialog(Postac, Tekst), !.

dialog_sad :-
    nl,
    dialog('Prokurator', 'Niech Pan opowowie jak z Pana perspektywy wyglądała rozmowa z szefem tego dnia.'),
    dialog('Andy Dufrasne', 'Był bardzo szorstki. Powiedział, że nikt już nie korzysta z vima, że zmniejsza on moją produktywność i zwolnienie mnie to jego jedyne wyjście.'),
    dialog('Prokurator', 'Czy po tym incydencie wypchnął Pan na maina gałąź, w której cały kod był pisany w jednej linii, a nazwy funkcji nie oddawały ich działania? Tu przykład: funkcja launch_missle() wypisywała na konsolę "Hello, World".'),
    dialog('Andy Dufrasne', 'Jak już mówiłem, zostałem wrobiony. Nigdy nie wypchnąłbym kodu przed pokryciu jego przynajmniej dziewięćdziesięciu procent testami jednostkowymi.'),
    write('...'), nl,
    dialog('Sędzia', 'Panie Andrew Dufrasne, za pisanie brzydkiego kodu zostaje Pan skazany na dwa dożywocia w więzieniu o zaostrzonym rygorze!'),
    nl, !.

ekwipunek :-
    findall(Przedmiot, ekwipunek(Przedmiot), Lista),
    ( Lista = [] ->
        ansi_format([fg(blue)], 'Twój ekwipunek jest pusty.~n', [])
    ;
        ansi_format([fg(blue)], 'Masz przy sobie:~n', []),
        forall(member(P, Lista), ansi_format([fg(blue)], '-~w~n', [P]))
    ),
    sprawdz_ucieczke, !.

idz(Miejsce) :-
    lokacja(Miejsce, Opis),
    retractall(pozycja_gracza(_)),
    assertz(pozycja_gracza(Miejsce)),
    ansi_format([fg(green)], '~w~n', [Opis]), !,
    rozejrzyj.

idz(_) :-
    ansi_format([fg(red)], 'Nie ma takiego miejsca!~n', []), !.

instrukcje :-
    nl,
    write('Dostępne komendy:'), nl,
    write('start.               -- rozpoczęcie gry.'), nl,
    % write('daj(imie, przedmiot) -- daj przedmiot z ekwipunku innej postaci.'), nl,
    write('ekwipunek.           -- sprawdź czy masz już wszystko do ucieczki.'), nl,
    write('idz(miejsce).        -- przejdź do "miejsce".'), nl,
    write('instrukcje.          -- wyświetl te komendy.'), nl,
    write('mapa.                -- wyświetl mapę więzienia.'), nl,
    write('porozmawiaj(imie).   -- rozpocznij rozmowę z postacią w tej lokacji'), nl,
    write('rozejrzyj.           -- rozejrzyj się dookoła.'), nl,
    write('uzyj(przedmiot).     -- spróbuj skorzystać z przedmiotu w ekwipunku'), nl,
    write('wez(przedmiot).      -- weź przedmiot z aktualnej lokacji.'), nl,
    write('halt.                -- zakończ rozgrywkę i wyjdź.'), nl,
    nl, !.

mapa :-
    nl,
    write('Miejsca, do których możesz przejść:'), nl,
    write('biblioteka           -- więzienna biblioteka'), nl,
    write('cela                 -- twoja cela'), nl,
    write('pralnia              -- pralnia, tu pracujesz'), nl,
    write('spacerniak           -- miejsce do spotkań z współwięźniami na dworze'), nl,
    write('stolowka             -- więzienna stołówka'), nl,
    nl, !.

porozmawiaj(red) :-
    \+ ekwipunek(kontakt),
    \+ ekwipunek(atlas),
    \+ przedmiot_w(atlas, biblioteka),
    pozycja_gracza(Miejsce),
    postac_w(red, Miejsce),
    dialog_postaci(red),
    assertz(przedmiot_w(atlas, biblioteka)), !.

porozmawiaj(red) :-
    ekwipunek(atlas),
    pozycja_gracza(spacerniak),
    dialog_postaci(red),
    retract(ekwipunek(atlas)),
    assertz(ekwipunek(kontakt)), !.

porozmawiaj(Rozmowca) :-
    pozycja_gracza(Miejsce),
    postac_w(Rozmowca, Miejsce),
    dialog_postaci(Rozmowca), !.

porozmawiaj(Rozmowca) :-
    ansi_format([fg(red)], 'Nie ma tu nikogo o imieniu "~w".~n', [Rozmowca]), !.

potrzebne_do_ucieczki([
    drut,
    farba,
    kontakt,
    material,
    mydlo,
    papier,    
    plaszcze,
    srubokret,
    sznurek,
    sztucce,
    ubrania,
    wlosy
]).

rozejrzyj :-
    pozycja_gracza(Miejsce),
    % przedmioty
    findall(Przedmiot, przedmiot_w(Przedmiot, Miejsce), Przedmioty),
    wypisz_przedmioty(Przedmioty),
    % postacie
    findall(Postac, postac_w(Postac, Miejsce), Postacie),
    wypisz_postacie(Postacie), !.

sprawdz_ucieczke :-
    potrzebne_do_ucieczki(Wymagane),
    ( forall(member(Rzecz, Wymagane), ekwipunek(Rzecz)) ->
        ucieczka_z_wiezienia
    ;
        true
    ).

start :-
    instrukcje,
    dialog_sad,
    start_przedmioty,
    start_postacie,
    start_rozmowy,
    idz(cela), !.

start_przedmioty :-
    retractall(przedmiot_w(_, _)),
    % assertz(przedmiot_w(atlas, biblioteka)),
    assertz(przedmiot_w(sztucce, stolowka)),
    assertz(przedmiot_w(ubrania, pralnia)),
    assertz(przedmiot_w(klej, biblioteka)),
    assertz(przedmiot_w(pamietnik, cela)),
    assertz(przedmiot_w(plaszcze, pralnia)),
    assertz(przedmiot_w(papier, biblioteka)),
    assertz(przedmiot_w(farba, spacerniak)),
    % assertz(przedmiot_w(nozyczki, biblioteka)),
    % assertz(przedmiot_w(jablko, stolowka)),
    assertz(przedmiot_w(odkurzacz, pralnia)),
    assertz(przedmiot_w(pizama, cela)),
    assertz(przedmiot_w(mydlo, cela)).

start_postacie :-
    retractall(postac_w(_, _)),
    assertz(postac_w(bibliotekarz, biblioteka)),
    assertz(postac_w(red, spacerniak)),
    assertz(postac_w(klawisz, pralnia)),
    assertz(postac_w(kucharz, stolowka)).

start_rozmowy :-
    retractall(kwestia(_, _)),
    assertz(kwestia('red', 'Czyli próbujesz uciec i potrzebujesz kogoś na zewnątrz do pomocy? Znajdź mi w bibliotece atlas, a w zamian zobaczę co da się zrobić.')),
    assertz(kwestia('bibliotekarz', 'Ciii... Próbuję się skupić na tej książcę o relacyjnych bazach danych.')),
    assertz(kwestia('klawisz', 'Nie masz dziś zmiany w pralni? Lepiej się pośpiesz!')),
    assertz(kwestia('kucharz', 'To co zwykle?')), !.

ucieczka_z_wiezienia :-
    nl,
    write('Masz wszystko czego potrzeba do ucieczki. Z mydła i skarpetek znalezionych w ubraniach robisz prowizoryczną broń. Pora poukrywać przedmioty po celi, żeby nie wzbudzały podejrzeń, poczekać na noc i rozpocząć ucieczkę.'), nl, nl,
    write('Otrzymujesz nowy zestaw instrukcji dostępny po wpisaniu polecenia "instrukcje".'), nl,
    reset_game, !.

uzyj(Przedmiot) :-
    ekwipunek(Przedmiot),
    akcja_uzycia(Przedmiot), !.

uzyj(Przedmiot) :-
    ansi_format([fg(red)], 'Nie masz albo nie możesz użyć przedmiotu: ~w~n', [Przedmiot]), !.

wez(atlas) :-
    pozycja_gracza(biblioteka),
    przedmiot_w(atlas, biblioteka),
    assertz(ekwipunek(atlas)),
    retract(przedmiot_w(atlas, biblioteka)),
    retractall(kwestia(red, _)),
    assertz(kwestia(red, 'Super, dzięki za atlas! Sprawę kontaktu możesz uznać za załatwioną.')), !.

wez(plaszcze) :-
    pozycja_gracza(pralnia),
    \+ przedmiot_w(jablko, stolowka),
    \+ ekwipunek(jablko),
    ansi_format([fg(red)], 'Klawisz patrzy! Musisz go czymś zająć, może jedzeniem?~n', []), !,
    assertz(przedmiot_w(jablko, stolowka)), !.

wez(plaszcze) :-
    pozycja_gracza(pralnia),
    ekwipunek(jablko),
    \+ ekwipunek(plaszcze),
    retract(przedmiot_w(plaszcze, pralnia)),
    retract(ekwipunek(jablko)),
    assertz(ekwipunek(plaszcze)),
    ansi_format([fg(green)], 'Zająłeś klawisza jabłkiem i zabrałeś płaszcze.~n', []), !.

wez(plaszcze) :-
    przedmiot_w(jablko, stolowka),
    ansi_format([fg(red)], 'Klawisz patrzy! Musisz go czymś zająć, może jedzeniem?~n', []), !.

wez(pizama) :-
    pozycja_gracza(cela),
    \+ przedmiot_w(nozyczki, biblioteka),
    \+ ekwipunek(nozyczki),
    assertz(przedmiot_w(nozyczki, biblioteka)),
    ansi_format([fg(red)], 'Bezużyteczna piżama... chyba że znajdziesz coś czym wytnie się z niej sznurek i materiał.~n', []), !.

wez(pizama) :-
    pozycja_gracza(cela),
    ekwipunek(nozyczki),
    retract(przedmiot_w(pizama, cela)),
    \+ ekwipunek(sznurek),
    assertz(ekwipunek(sznurek)),
    assertz(ekwipunek(material)),
    ansi_format([fg(green)], 'Wyciąłeś z piżamy sznurek i materiał.~n', []), !.

wez(pizama) :-
    przedmiot_w(nozyczki, biblioteka),
    ansi_format([fg(red)], 'Bezużyteczna piżama... chyba że znajdziesz coś czym wytnie się z niej sznurek i materiał.~n', []), !.

wez(odkurzacz) :-
    pozycja_gracza(pralnia),
    \+ przedmiot_w(srubokret, spacerniak),
    \+ ekwipunek(srubokret),
    assertz(przedmiot_w(srubokret, spacerniak)),
    ansi_format([fg(red)], 'Odkurzacz. Przydałoby się jakoś go rozkręcić, żeby dostać się do wnętrzności...~n', []), !.
    
wez(odkurzacz) :-
    pozycja_gracza(pralnia),
    ekwipunek(srubokret),
    \+ ekwipunek(drut),
    assertz(ekwipunek(drut)),
    retract(przedmiot_w(odkurzacz, pralnia)),
    ansi_format([fg(green)], 'Rozkręciłeś odkurzacz i wyjąłeś z niego drut.~n', []), !.

wez(odkurzacz) :-
    przedmiot_w(srubokret, spacerniak),
    ansi_format([fg(green)], 'Znajdź coś do rozkręcenia odkurzacza.~n', []), !.

wez(Przedmiot) :-
    pozycja_gracza(Miejsce),
    przedmiot_w(Przedmiot, Miejsce),
    retract(przedmiot_w(Przedmiot, Miejsce)),
    assertz(ekwipunek(Przedmiot)),
    ansi_format([fg(green)], 'Zabrałeś: ~w~n', [Przedmiot]), !.

wypisz_postacie([]).

wypisz_postacie(Postacie) :-
    ansi_format([fg(cyan)], 'Spotykasz tutaj:~n', []),
    forall(member(P,Postacie), ansi_format([fg(cyan)], '-~w~n', [P])), !.

wypisz_przedmioty([]) :-
    ansi_format([fg(blue)], 'Nie ma tu nic ciekawego.~n', []), !.

wypisz_przedmioty(Przedmioty) :-
    ansi_format([fg(magenta)], 'Widzisz:~n', []),
    forall(member(P, Przedmioty), ansi_format([fg(magenta)], '-~w~n', [P])), !.



reset_game :-
    retractall(ekwipunek(_)),
    retractall(kwestia(_, _)),
    retractall(postac_w(_, _)),
    retractall(pozycja_gracza(_)),
    retractall(przedmiot_w(_, _)),

    abolish(ekwipunek/1),
    abolish(kwestia/2),
    abolish(postac_w/2),
    abolish(pozycja_gracza/1),
    abolish(przedmiot_w/2),
    abolish(akcja_uzycia/1),
    abolish(daj/2),
    abolish(dialog/2),
    abolish(dialog_postaci/1),
    abolish(dialog_sad/0),
    abolish(ekwipunek/0),
    abolish(idz/1),
    abolish(instrukcje/0),
    abolish(mapa/0),
    abolish(porozmawiaj/1),
    abolish(potrzebne_do_ucieczki/1),
    abolish(rozejrzyj/0),
    abolish(sprawdz_ucieczke/0),
    abolish(start/0),
    abolish(start_przedmioty/0),
    abolish(start_postacie/0),
    abolish(start_rozmowy/0),
    abolish(ucieczka_z_wiezienia/0),
    abolish(uzyj/1),
    abolish(wez/1),
    abolish(wypisz_postacie/1),
    abolish(wypisz_przedmioty/1),

    consult('cell_escape.pl').

File: haskell/app/prison_escape.hs
{-# LANGUAGE RecordWildCards #-}

import Control.Monad.State
import Control.Monad (unless)
import Control.Monad (when)
import System.Console.ANSI
import qualified Data.Map as Map
import Data.Maybe (fromMaybe)
import Data.List (intercalate)


data Miejsce = Biblioteka | Cela | Pralnia | Spacerniak | Stolowka
    deriving (Show, Eq, Ord, Enum)


data Przedmiot = Atlas | Drut | Farba | Jablko | Klej | Kontakt | Material | Mydlo | Nozyczki | Odkurzacz | Pamietnik | Papier | Pizama | Plaszcze | Sztucce | Sznurek | Srubokret | Ubrania | Wlosy
    deriving (Show, Eq, Ord)


data Postac = Redding | Bibliotekarz | Klawisz | Kucharz
    deriving (Show, Eq, Ord)


data StanGry = StanGry
    { miejsceGracza :: Miejsce
    , ekwipunek :: [Przedmiot]
    , przedmiotyLokacji :: Map.Map Miejsce [Przedmiot]
    , postacieLokacji :: Map.Map Miejsce [Postac]
    , dialogi :: Map.Map Postac String
    , przedmiotyDoUcieczki :: [Przedmiot]
    }


stanPoczatkowy :: StanGry
stanPoczatkowy = StanGry
    { miejsceGracza = Cela
    , ekwipunek = []
    , przedmiotyLokacji = Map.fromList
        [ (Biblioteka, [Klej, Papier])
        , (Cela, [Mydlo, Pamietnik, Pizama])
        , (Pralnia, [Odkurzacz, Plaszcze, Ubrania])
        , (Spacerniak, [Farba])
        , (Stolowka, [Sztucce])
        ]
    , postacieLokacji = Map.fromList
        [ (Biblioteka, [Bibliotekarz])
        , (Pralnia, [Klawisz])
        , (Spacerniak, [Redding])
        , (Stolowka, [Kucharz])
        ]
    , dialogi = Map.fromList
        [ (Bibliotekarz, "Ciii... Próbuję się skupić na tej książcę o relacyjnych bazach danych.")
        , (Klawisz, "Nie masz dziś zmiany w pralni? Lepiej się pośpiesz!")
        , (Kucharz, "To samo co zwykle?")
        , (Redding, "Czyli próbujesz uciec i potrzebujesz kogoś na zewnątrz do pomocy? Znajdź mi w bibliotece atlas, a w zamian zobaczę co da się zrobić.")
        ]
    , przedmiotyDoUcieczki = [Drut, Farba, Kontakt, Material, Mydlo, Papier, Plaszcze, Srubokret, Sznurek, Sztucce, Ubrania, Wlosy]
    }


type Gra a = StateT StanGry IO a


wypiszKolor :: Color -> String -> IO ()
wypiszKolor kolor tekst = do
    setSGR [SetColor Foreground Vivid kolor]
    putStrLn tekst
    setSGR [Reset]


opisMiejsca :: Miejsce -> String
opisMiejsca Biblioteka = "Jesteś w bibliotece."
opisMiejsca Cela = "Jesteś w swojej celi."
opisMiejsca Pralnia = "Jesteś w pralni."
opisMiejsca Spacerniak = "Jesteś na spacerniaku."
opisMiejsca Stolowka = "Jesteś na więziennej stołówce."


idz :: Miejsce -> Gra ()
idz miejsce = do
    state <- get
    put state {miejsceGracza = miejsce}
    liftIO $ wypiszKolor Green (opisMiejsca miejsce)
    rozejrzyj


rozejrzyj :: Gra ()
rozejrzyj = do
    StanGry {..} <- get
    let przedmioty = fromMaybe [] (Map.lookup miejsceGracza przedmiotyLokacji)
        postacie = fromMaybe [] (Map.lookup miejsceGracza postacieLokacji)
    liftIO $ do
        if null przedmioty 
            then wypiszKolor Blue "Nie ma tu nic ciekawego."
            else do
                wypiszKolor Magenta "Widzisz: "
                mapM_ (\przedmiot -> wypiszKolor Magenta $ "- " ++ show przedmiot) przedmioty
        if not (null postacie)
            then do
                wypiszKolor Cyan "Spotykasz tutaj: "
                mapM_ (\postac -> wypiszKolor Cyan $ "- " ++ show postac) postacie
            else return ()


wez :: Przedmiot -> Gra ()
wez przedmiot = do
    state@StanGry {..} <- get
    let przedmiotyTutaj = fromMaybe [] (Map.lookup miejsceGracza przedmiotyLokacji)
    if przedmiot `elem` przedmiotyTutaj
        then case (przedmiot, miejsceGracza) of
            -- przypadek dla pizamy
            (Pizama, Cela) | Nozyczki `notElem` ekwipunek && not (Nozyczki `elem` fromMaybe [] (Map.lookup Biblioteka przedmiotyLokacji)) -> do
                let nowePrzedmiotyLokacji = Map.adjust (Nozyczki :) Biblioteka przedmiotyLokacji 
                put state { przedmiotyLokacji = nowePrzedmiotyLokacji}
                liftIO $ wypiszKolor Red "Bezużyteczna piżama... chyba że znajdziesz coś czym wytnie się z niej sznurek i materiał."
            (Pizama, Cela) | Nozyczki `elem` ekwipunek && Sznurek `notElem` ekwipunek && Material `notElem` ekwipunek -> do
                let nowePrzedmiotyTutaj = filter (/= Pizama) przedmiotyTutaj
                    nowePrzedmiotyMapa = Map.insert miejsceGracza nowePrzedmiotyTutaj przedmiotyLokacji
                put state { ekwipunek = Sznurek : Material : ekwipunek, przedmiotyLokacji = nowePrzedmiotyMapa }
                liftIO $ wypiszKolor Green "Wyciąłeś z piżamy sznurek i materiał."
            (Pizama, Cela) | Nozyczki `elem` fromMaybe [] (Map.lookup Biblioteka przedmiotyLokacji) -> do
                liftIO $ wypiszKolor Red "Bezużyteczna piżama... chyba że znajdziesz coś czym wytnie się z niej sznurek i materiał."

            -- przypadek dla odkurzacza
            (Odkurzacz, Pralnia) | Srubokret `notElem` ekwipunek && not (Srubokret `elem` fromMaybe [] (Map.lookup Spacerniak przedmiotyLokacji)) -> do
                let nowePrzedmiotyLokacji = Map.adjust (Srubokret :) Spacerniak przedmiotyLokacji
                put state { przedmiotyLokacji = nowePrzedmiotyLokacji }
                liftIO $ wypiszKolor Red "Odkurzacz. Przydałoby się jakoś go rozkręcić, żeby dostać się do wnętrzności..."
            (Odkurzacz, Pralnia) | Srubokret `elem` ekwipunek && Drut `notElem` ekwipunek -> do
                let nowePrzedmiotyTutaj = filter (/= Odkurzacz) przedmiotyTutaj
                    nowePrzedmiotyMapa = Map.insert miejsceGracza nowePrzedmiotyTutaj przedmiotyLokacji
                put state { ekwipunek = Drut : ekwipunek, przedmiotyLokacji = nowePrzedmiotyMapa }
                liftIO $ wypiszKolor Green "Rozkręciłeś odkurzacz i wyjąłeś z niego drut."
            (Odkurzacz, Pralnia) | Srubokret `elem` fromMaybe [] (Map.lookup Spacerniak przedmiotyLokacji) -> do
                liftIO $ wypiszKolor Red "Odkurzacz. Przydałoby się jakoś go rozkręcić, żeby dostać się do wnętrzności..."

            -- przypadek dla plaszczy
            (Plaszcze, Pralnia) | Jablko `notElem` ekwipunek && not (Jablko `elem` fromMaybe [] (Map.lookup Stolowka przedmiotyLokacji)) -> do
                let nowePrzedmiotyLokacji = Map.adjust (Jablko :) Stolowka przedmiotyLokacji
                put state { przedmiotyLokacji = nowePrzedmiotyLokacji }
                liftIO $ wypiszKolor Red "Klawisz patrzy! Musisz go czymś zająć, może jedzeniem?"
            (Plaszcze, Pralnia) | Jablko `elem` ekwipunek && Plaszcze `notElem` ekwipunek -> do
                let nowePrzedmiotyTutaj = filter (/= Plaszcze) przedmiotyTutaj
                    nowePrzedimotyMapa = Map.insert miejsceGracza nowePrzedmiotyTutaj przedmiotyLokacji
                put state { ekwipunek = Plaszcze : filter (/= Jablko) ekwipunek, przedmiotyLokacji = nowePrzedimotyMapa }
                liftIO $ wypiszKolor Green "Zająłeś klawisza jabłkiem i zabrałeś płaszcze."
            (Plaszcze, Pralnia) | Jablko `elem` fromMaybe [] (Map.lookup Stolowka przedmiotyLokacji) -> do
                liftIO $ wypiszKolor Red "Klawisz patrzy! Musisz go czymś zająć, może jedzeniem?"

            -- przypadek dla atlasu
            (Atlas, Biblioteka) -> do
                let nowePrzedmiotyTutaj = filter (/= Atlas) przedmiotyTutaj
                    nowePrzedimotyMapa = Map.insert miejsceGracza nowePrzedmiotyTutaj przedmiotyLokacji
                    noweDialogi = Map.insert Redding "Super, dzięki za atlas! Sprawę kontaktu możesz uznać za załatwioną." dialogi 
                put state { ekwipunek = Kontakt : Atlas : ekwipunek, przedmiotyLokacji = nowePrzedimotyMapa, dialogi = noweDialogi }
                liftIO $ wypiszKolor Green $ "Zabrałeś: " ++ show przedmiot

            -- domyslny przypadek
            _ -> do
                let nowePrzedmiotyTutaj = filter (/= przedmiot) przedmiotyTutaj
                    nowePrzedimotyMapa = Map.insert miejsceGracza nowePrzedmiotyTutaj przedmiotyLokacji
                put state { ekwipunek = przedmiot : ekwipunek, przedmiotyLokacji = nowePrzedimotyMapa }
                liftIO $ wypiszKolor Green $ "Zabrałeś: " ++ show przedmiot
            else liftIO $ wypiszKolor Red $ "Nie ma tu przedmiotu: " ++ show przedmiot


uzyj :: Przedmiot -> Gra ()
uzyj przedmiot = do
    state@StanGry {..} <- get
    if przedmiot `elem` ekwipunek
        then case przedmiot of
            Pamietnik -> do
                liftIO $ do
                    wypiszKolor White $ "Potrzebne do ucieczki:"
                    mapM_ (\przedmiot -> wypiszKolor White $ "- " ++ show przedmiot) przedmiotyDoUcieczki
            Nozyczki -> if miejsceGracza == Cela && Wlosy `notElem` ekwipunek
                then do
                    put state {ekwipunek = Wlosy : ekwipunek}
                    liftIO $ wypiszKolor Green "Obciąłeś włosy."
                else liftIO $ wypiszKolor Red "Chcesz obciąć włosy? Mądre... ale zrób to w celi, żeby nikt się nie zainteresował."
            _ -> liftIO $ wypiszKolor Red $ "Nie możesz użyć przedmiotu: " ++ show przedmiot
    else liftIO $ wypiszKolor Red $ "Nie masz przedmiotu: " ++ show przedmiot


porozmawiaj :: Postac -> Gra ()
porozmawiaj postac = do
    state@StanGry {..} <- get
    let postacieTutaj = fromMaybe [] (Map.lookup miejsceGracza postacieLokacji)
    if postac `elem` postacieTutaj
        then do
            liftIO $ do
                wypiszDialog (show postac) $ fromMaybe "" (Map.lookup postac dialogi)
            when (postac == Redding && Kontakt `notElem` ekwipunek && Atlas `notElem` ekwipunek && not (Atlas `elem` fromMaybe [] (Map.lookup Biblioteka przedmiotyLokacji))) $ do
                let nowePrzedmiotyLokacji = Map.adjust (Atlas :) Biblioteka przedmiotyLokacji
                put state { przedmiotyLokacji = nowePrzedmiotyLokacji}
        else liftIO $ wypiszKolor Red $ "Nie ma tu nikogo o imieniu: " ++ show postac


pokazEkwipunek :: Gra ()
pokazEkwipunek = do
    StanGry {..} <- get
    liftIO $ if null ekwipunek
        then wypiszKolor Blue "Twój ekwipunek jest pusty."
        else do
            wypiszKolor Blue "Masz przy sobie:"
            mapM_ (\przedmiot -> wypiszKolor Blue $ "- " ++ show przedmiot) ekwipunek
    sprawdzUcieczke


sprawdzUcieczke :: Gra ()
sprawdzUcieczke = do
    StanGry {..} <- get
    if all (`elem` ekwipunek) przedmiotyDoUcieczki
        then liftIO $ do
            wypiszKolor White "Masz wszystko czego potrzeba do ucieczki. Z mydła i skarpetek znalezionych w ubraniach robisz prowizoryczną broń. Pora poukrywać prz    edmioty po celi, żeby nie wzbudzały podejrzeń, poczekać na noc i rozpocząć ucieczkę."
            wypiszKolor White "Otrzymujesz nowy zestaw instrukcji dostępny po wpisaniu polecenia 'instrukcje'."
        else return ()


wypiszDialogStartowy :: IO ()
wypiszDialogStartowy = do
    let dialogStartowy = 
            [ ("Prokurator", "Niech Pan opowie jak z Pana perspektywy wyglądała rozmowa z szefem tego dnia.")
            , ("Andy Dufresne", "Był bardzo szorstki. Powiedział, że nikt już nie korzysta z vima, że zmniejsza on moją produktywność i zwolnienie mnie to jego jedyne wyjście.")
            , ("Prokurator", "Czy po tym incydencie wypchnął Pan na maina gałąź, w której cały kod był pisany w jednej linii, a nazwy funkcji nie oddawały ich działania? Tu przykład: funkcja launch_missle() wypisywała na konsolę 'Hello, World'.")
            , ("Andy Dufrasne", "Jak już mówiłem, zostałem wrobiony. Nigdy nie wypchnąłbym kodu przed pokryciu jego przynajmniej dziewięćdziesięciu procent testami jednostkowymi.")
            , ("Sędzia", "Panie Andrew Dufrasne, za pisanie brzydkiego kodu zostaje Pan skazany na dwa dożywocia w więzieniu o zaostrzonym rygorze!")
            ]
    mapM_ (uncurry wypiszDialog) dialogStartowy


wypiszDialog :: String -> String -> IO ()
wypiszDialog postac tekst = do
    setSGR [SetColor Foreground Vivid Cyan]
    putStr postac
    putStr ": "
    setSGR [SetColor Foreground Vivid White]
    putStrLn tekst
    setSGR [Reset]


instrukcje :: Gra ()
instrukcje = liftIO $ do
    putStrLn "Dostępne komendy:"
    putStrLn "start                 -- rozpoczęcie gry."
    putStrLn "ekwipunek             -- sprawdź czy masz już wszystko do ucieczki."
    putStrLn "idz <miejsce>         -- przejdź do miejsca."
    putStrLn "instrukcje            -- wyświetl te komendy."
    putStrLn "mapa                  -- wyświetl mapę więzienia."
    putStrLn "porozmawiaj <imie>    -- rozpocznij rozmowę z postacią w tej lokacji."
    putStrLn "rozejrzyj             -- rozejrzyj się dookoła."
    putStrLn "uzyj <przedmiot>      -- spróbuj skorzystać z przedmiotu w ekwipunku."
    putStrLn "wez <przedmiot>       -- weź przedmiot z aktualnej lokacji."
    putStrLn "wyjdz                 -- zakończ rozgrywkę i wyjdź."


mapa :: Gra ()
mapa = liftIO $ do
  putStrLn "Miejsca, do których możesz przejść:"
  putStrLn "biblioteka           -- więzienna biblioteka"
  putStrLn "cela                 -- twoja cela"
  putStrLn "pralnia              -- pralnia, tu pracujesz"
  putStrLn "spacerniak           -- miejsce do spotkań z współwięźniami na dworze"
  putStrLn "stolowka             -- więzienna stołówka"


petlaGry :: Gra ()
petlaGry = do
    liftIO $ putStr "> "
    cmd <- liftIO getLine
    case words cmd of
        ["ekwipunek"] -> pokazEkwipunek
        ["idz", miejsce] -> case miejsce of
            "biblioteka" -> idz Biblioteka
            "cela" -> idz Cela
            "pralnia" -> idz Pralnia
            "spacerniak" -> idz Spacerniak
            "stolowka" -> idz Stolowka
            _ -> liftIO $ wypiszKolor Red "Nie ma takiego miejsca!"
        ["instrukcje"] -> do instrukcje 
        ["mapa"] -> mapa
        ["porozmawiaj", postac] -> case postac of
            "bibliotekarz" -> porozmawiaj Bibliotekarz
            "klawisz" -> porozmawiaj Klawisz
            "kucharz" -> porozmawiaj Kucharz 
            "redding" -> porozmawiaj Redding 
            _ -> liftIO $ wypiszKolor Red $ "Nie ma tu nikogo o imieniu " ++ show postac
        ["rozejrzyj"] -> rozejrzyj
        ["uzyj", przedmiot] -> case przedmiot of
            "pamietnik" -> uzyj Pamietnik
            "nozyczki" -> uzyj Nozyczki
            _ -> liftIO $ wypiszKolor Red $ "Nie można użyć przedmiotu: " ++ show przedmiot
        ["wez", przedmiot] -> case przedmiot of
            "atlas" -> wez Atlas
            "sztucce" -> wez Sztucce
            "ubrania" -> wez Ubrania
            "klej" -> wez Klej
            "pamietnik" -> wez Pamietnik
            "plaszcze" -> wez Plaszcze
            "papier" -> wez Papier
            "farba" -> wez Farba
            "nozyczki" -> wez Nozyczki
            "jablko" -> wez Jablko
            "odkurzacz" -> wez Odkurzacz
            "pizama" -> wez Pizama
            "mydlo" -> wez Mydlo
            "srubokret" -> wez Srubokret
            _ -> liftIO $ wypiszKolor Red $ "Nie ma tu przedmiotu: " ++ przedmiot 
        ["wyjdz"] -> return ()
        _ -> liftIO $ wypiszKolor Red "Nieznana komenda!"
    unless (cmd == "wyjdz") petlaGry


main :: IO ()
main = do
    wypiszDialogStartowy
    evalStateT (instrukcje >> idz Cela >> petlaGry) stanPoczatkowy
